wdio-allure-reporter: Add capability to include attachment in afterhook

PippoRaimondi opened the pull request:
## Proposed changes

Adding a simple flag to a reporter to allow us to add an existing attachment in the after hook

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### Reviewers: @webdriverio/technical-committee

christian-bromann commented on packages/wdio-allure-reporter/src/index.js:
```diff
@@ -249,9 +249,11 @@ class AllureReporter extends WDIOReporter {
-        if (!this.isAnyTestRunning()) {
-            return false
+    addAttachment({ name, content, type = 'text/plain' }, loadExistingAttachmentInAfterHook = false) {
+        if (!loadExistingAttachmentInAfterHook) {
```
```suggestion
        if (!loadExistingAttachmentInAfterHook || !this.isAnyTestRunning()) {
```
no need for two nested ifs

PippoRaimondi replied:
No, we need to ensure the nested condition is not interpreted at all if the flag is true...  it's logically incorrect to use `||` or `&&` in this case.

BorisOsipov commented:
>Adding a simple flag to a reporter to allow us to add an existing attachment in the after hook

can you describe your which issue you solve? For me it is not clear why we need some boolean flag. Looks like this condition can be easy implemented in your own code..

PippoRaimondi commented:
> > Adding a simple flag to a reporter to allow us to add an existing attachment in the after hook
> 
> can you describe your which issue you solve? For me it is not clear why we need some boolean flag. Looks like this condition can be easy implemented in your own code..

I am going to use this report to attach screenshots from my visual regression test suite so it's easier to debug in case something goes wrong. And so I don't use the `addAttachment` method in each and every test rather than in a global hook, I need to ensure it doesn't return... does this make sense?

christian-bromann commented:
@PippoRaimondi I agree with @BorisOsipov you could just instead of calling
```js
const loadExistingAttachmentInAfterHook = false
// ...
addAttachment({ ... }, loadExistingAttachmentInAfterHook)
```
do
```js
if (loadExistingAttachmentInAfterHook) {
    addAttachment({ ... })
```

PippoRaimondi commented:
> @PippoRaimondi I agree with @BorisOsipov you could just instead of calling
> 
> ```js
> const loadExistingAttachmentInAfterHook = false
> // ...
> addAttachment({ ... }, loadExistingAttachmentInAfterHook)
> ```
> 
> do
> 
> ```js
> if (loadExistingAttachmentInAfterHook) {
>     addAttachment({ ... })
> }
> ```

Please note the wdio allure reporter instance method `addAttachment` is a wrapper to allures method `addAttachment`, which is what I am trying to get invoked here... The below condition prevents allures instance `addAttachment` from being called.

```
    if (!this.isAnyTestRunning()) {
        return false
    }
```

In the `afterEach` hook `this.isAnyTestRunning()` returns `false`... hence I needed a way to bypass that so I could attach my screenshots... I don't have a lot of context why we added that check first place, so with the little information I had I added a flag to ensure we didn't break any existing functionality...

christian-bromann commented:
> In the `afterEach` hook

Which afterEach hook? The one in your code or in one of the plugins?

PippoRaimondi commented:
> > In the `afterEach` hook
> 
> Which afterEach hook? The one in your code or in one of the plugins?

In my code, within wdio test suite...

christian-bromann commented:
> In my code, within wdio test suite

So then add an additional `if` in your test suite as I mentioned in my comment above.

PippoRaimondi commented:
> > In my code, within wdio test suite
> 
> So then add an additional `if` in your test suite as I mentioned in my comment above.

Sorry, I might have not explained it very well the reason why it won't work if I add an `if` in my test suite.

Let me test something out (might add a new method all together) then I will come back to you with a more detailed reason why I am doing it in the package.

PippoRaimondi commented:
> > > In my code, within wdio test suite
> > 
> > 
> > So then add an additional `if` in your test suite as I mentioned in my comment above.
> 
> Sorry, I might have not explained it very well the reason why it won't work if I add an `if` in my test suite.
> 
> Let me test something out (might add a new method all together) then I will come back to you with a more detailed reason why I am doing it in the package.

Ok, so this will never work by changing `wdio-allure-reporter` package alone.

Quick recap, the flag I was adding was attempting to bypass condition that returns `false` from the method below. This was one of the reasons why calling the addAttachment out of the test context wouldn't work, i.e outside of the `it` block... What I was trying to avoid was have to call `addAttachment` in each test (within the `it` block)...

```
addAttachment({ name, content, type = 'text/plain' }) {
    **if (!this.isAnyTestRunning()) {
        return false
    }**

    if (type === 'application/json') {
        this.dumpJSON(name, content)
    } else {
        this.allure.addAttachment(name, Buffer.from(content), type)
    }
}
```

Although it was a bit sloppy at my end thinking the condition in bold above was the only thing needed to be changed (also the update was incomplete, I had to change the static reference as well to actually get the flag to "work")...

I looked deeper and the real blocker is with `allure-js-common`... by the end of the test or suite it calls `endCase` or `endSuite` and unless the architecture of that bit is changed the methods can never really be called in hooks as they lose all the test context... To fix this would require a bit of work and not sure there is enough value as I have a workaround to this issue.

Unless you have any other idea of how to accomplish this, it's safe to close this PR.

christian-bromann commented:
Thanks for looking into it. Closing!
